src/main.ts

```diff
@@ -0,0 +1,142 @@
+
+import config from './config'
+
+const confFile = ".atom/ide-reason.json"
+const defaultConfig = languageServer.ISettings.defaults.reason
+type Config = languageServer.ISettings['reason']
+
+const scopes = [
+  'source.reason',
+  'source.ocaml',
+  'source.re',
+  'source.ml',
+  'source.rei',
+  'source.mli',
+]
+
+function flatten1<T>(array: T[]) {
+  return ([] as T[]).concat(...array)
+}
+
+function joinEnvPath(...paths: (string | undefined)[]) {
+  return flatten1(paths)
+    .filter(Boolean)
+    .join(path.delimiter)
+}
+
+function readFileConf(dir: string, file: string, property?: string) {
+  try {
+    const conf = require(path.join(dir, file))
+    return property ? conf[property] : conf
+  } catch (e) {
+    console.error(e)
+  }
+  return null
+}
+
+class ReasonMLLanguageClient extends AutoLanguageClient {
+  subscriptions: CompositeDisposable | null = null
+  confPerProject: Partial<Config> = {}
+
+  getGrammarScopes () { return scopes }
+  getLanguageName () { return 'Reason' }
+  getServerName () { return 'ocamlmerlin' }
+  getConnectionType() { return 'ipc' }
+  getRootConfigurationKey() { return 'ide-reason' }
+
+  mapConfigurationObject(config: Config) {
+    config = merge({}, config, this.confPerProject)
+    return {
+      reason: {
+        codelens: {
+          enabled: false,
+          unicode: true,
+        },
+        debounce: config.debounce,
+        diagnostics: {
+          tools: config.diagnostics.tools,
+          merlinPerfLogging: false,
+        },
+        format: config.format,
+        path: config.path,
+        server: config.server,
+      },
+    }
+  }
+
+  activate() {
+    super.activate()
+    require('atom-package-deps').install('ide-reason')
+      .then(() => {
+        console.log('All dependencies installed, good to go')
+      })
+
+    this.subscriptions = new CompositeDisposable();
+
+    this.subscriptions.add(atom.commands.add('atom-workspace', {
+      [`${pkg.name}:generateConfig`]: () => this.generateConfig().catch(console.error)
+    }));
+  }
+
+  async generateConfig() {
+    const paths = atom.project.getPaths()
+    let curProject = paths[0]
+    const editor = atom.workspace.getActiveTextEditor()
+    if (editor) {
+      const editorPath = editor.getPath()
+      if (editorPath) {
+        curProject = paths.filter(p => editorPath.startsWith(p))[0] || curProject
+      }
+    }
+    const confPath = path.join(curProject, confFile)
+    if (fs.existsSync(confPath)) {
+      return atom.notifications.addWarning('ide-reason.json already exists!', {
+        description: confPath,
+      })
+    }
+    await fs.outputFile(
+      confPath,
+      JSON.stringify(defaultConfig, null, 2),
+      'utf-8',
+    )
+  }
+
+  startServerProcess(projectPath: string) {
+    const serverPath = require.resolve('ocaml-language-server/bin/server')
+
+
+    const confFromPkg: Partial<Config> = readFileConf(projectPath, 'package.json', 'atom-reason') || {}
```

Comment:
Is it really Partial<Config>? I think RecursivePartial is more accurate. see also https://stackoverflow.com/questions/41980195/recursive-partialt-in-typescript-2-1

Reply:
Nice trick, but it's not working with deepmerge's tsd in this case.

Reply:
types/deepmerge has bugs about ts's default export, I'll take your advice!